core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java

```diff
@@ -32,54 +33,102 @@
+        this.clusterService = clusterService;
+        this.searchAction = searchAction;
+    }
+
     @Override
-    protected void doExecute(final MultiSearchRequest request, final ActionListener<MultiSearchResponse> listener) {
+    protected void doExecute(MultiSearchRequest request, ActionListener<MultiSearchResponse> listener) {
         ClusterState clusterState = clusterService.state();
         clusterState.blocks().globalBlockedRaiseException(ClusterBlockLevel.READ);
 
+        int maxParallelSearches = request.maxParallelSearchRequests();
         final AtomicArray<MultiSearchResponse.Item> responses = new AtomicArray<>(request.requests().size());
-        final AtomicInteger counter = new AtomicInteger(responses.length());
-        for (int i = 0; i < responses.length(); i++) {
-            final int index = i;
-            searchAction.execute(request.requests().get(i), new ActionListener<SearchResponse>() {
+        SearchRequestIterator iterator = new SearchRequestIterator(request.requests());
+        batchExecute(iterator, responses, maxParallelSearches, listener);
+    }
+
+    void batchExecute(SearchRequestIterator requests, AtomicArray<MultiSearchResponse.Item> responses,
+                      int maxParallelSearches, ActionListener<MultiSearchResponse> listener) {
+        int numParallelSearches = Math.min(requests.remaining(), maxParallelSearches);
+        final AtomicInteger counter = new AtomicInteger(numParallelSearches);
+        for (int i = 0; i < numParallelSearches; i++) {
+            int pos = requests.upto;
+            SearchRequest request = requests.next();
+            searchAction.execute(request, new ActionListener<SearchResponse>() {
                 @Override
                 public void onResponse(SearchResponse searchResponse) {
-                    responses.set(index, new MultiSearchResponse.Item(searchResponse, null));
-                    if (counter.decrementAndGet() == 0) {
-                        finishHim();
-                    }
+                    responses.set(pos, new MultiSearchResponse.Item(searchResponse, null));
+                    nextBatch();
                 }
 
                 @Override
                 public void onFailure(Throwable e) {
-                    responses.set(index, new MultiSearchResponse.Item(null, e));
-                    if (counter.decrementAndGet() == 0) {
-                        finishHim();
-                    }
+                    responses.set(pos, new MultiSearchResponse.Item(null, e));
+                    nextBatch();
                 }
 
-                private void finishHim() {
-                    listener.onResponse(new MultiSearchResponse(responses.toArray(new MultiSearchResponse.Item[responses.length()])));
+                private void nextBatch() {
+                    if (counter.decrementAndGet() == 0) {
```

Comment:
I think that we can do better than this. If I'm reading this correctly, this means that we wait until an entire batch is complete before submitting another batch of requests. Thus, a slow request can hold the next batch and thus the response. I think instead we should try to maintain as many requests in the queue as possible, up to the concurrent request limit.